### ISSUE TITLE
productionlist and token do not pass their IDs through nodes.make_id

### ISSUE THREAD

**Issue by cpitclaudel** [sphinx-doc/sphinx#4983]
Take this example from the manual:

```rst
.. productionlist::
   try_stmt: try1_stmt | try2_stmt
   try1_stmt: "try" ":" `suite`
            : ("except" [`expression` ["," `target`]] ":" `suite`)+
            : ["else" ":" `suite`]
            : ["finally" ":" `suite`]
   try2_stmt: "try" ":" `suite`
            : "finally" ":" `suite`
```

The generated HTML looks like this:

```html
…<strong id="grammar-token-try_stmt">try_stmt </strong> ::=  try1_stmt | try2_stmt…
```

Note how the ID contains an `_` character.  The responsible code is here:

https://github.com/sphinx-doc/sphinx/blob/4cdb51be8385baaa81ae815e2e84f1f81c934adf/sphinx/domains/std.py#L435

For comparison, here's how index entries are generated:

https://github.com/sphinx-doc/sphinx/blob/4cdb51be8385baaa81ae815e2e84f1f81c934adf/sphinx/domains/std.py#L256

Is it an oversight?  Should the code call `nodes.make_id`?

To clarify: I don't think it's a big issue, except when production names contain symbols, like this:

```rst
.. productionlist::
   except#name : except `name`
   except#expr : except (`expr`)

:token:`except#name`
```

which generates this `<strong id="grammar-token-except#name">except#name</strong>` and `<a class="reference internal" href="#grammar-token-except#name">…</a>`.  It's also not obvious how to fix this, since a fix might break links pointing to existing documentation.

**Comment by tk0miya** [comment]
I think wrong IDs will cause some problem. So they should be fixed.
I know that will break some kind of URLs.  So it is a kind of breaking change.

**Comment by tk0miya** [comment]
Fixed by #5000.
Thank you for reporting

**Comment by cpitclaudel** [comment]
Thanks for the fix!
